design/service-apis-implementation.md

```diff
@@ -0,0 +1,342 @@
+Contour will not update the `status.addresses` field.
+
+For Contour, the key Gateway section are the Listeners. These define how an implementation should listen for traffic described by Routes.
+
+(The remainder of this document assumes that [#3263](https://github.com/projectcontour/contour/pull/3263) has been approved and merged.)
+
+For Listeners:
+- Listeners inside Gateways will be merged where possible.
+- Listeners that specify a port that the Contour controller considers the default secure port will have a HTTP->HTTPS redirect created for them.
+- Listeners that specify a port that is not exposed via the Envoy service will be rejected.
+- Conflicts within a Gateway will result in the relevant Listeners both being rejected (as there is no way to determine which one was first).
+- Listeners are considered mergeable if all the fields out of `hostname`, `port`, and `protocol` match, with some additional rules around TLS.
+- Further merging rules are specified in the detailed design below.
+- Listeners that refer to any other Route than HTTPRoute or TLSRoute will be ignored, and a condition placed on the corresponding `status.listeners[]` object saying that it was ignored because those objects are not supported.
+
+
+The Gateway may supply TLS config, in which case it is used as a default.
+The TLS config may be overridden by HTTPRoutes if the Gateway `spec.listeners[].tls.routeOverride` is set to `Allow`.
+This allows the Gateway to configure a default TLS certificate.
+Note that this feature has fiddly interactions with the various places in which a hostname may be specified;
+this will require careful test case design.
+
+The output of this watcher is both Envoy configuration, and a list of kind/name/namespace details for Routes to watch.
+
+Contour will only update the `status` of Gateway objects.
+
+### HTTPRoute
+Contour will watch all HTTPRoute objects, and filter for entries matching a label selector in its configured Gateway's `spec.listeners[].routes`,
+and will configure the associated routes.
+Configuration of HTTPRoutes will be subject to the rules around the `RouteGateways` field.
+
+When a HTTPRoute specifies a hostname or slice of hostnames, those hostnames must match the hostnames in the Gateway.
+Note that more specific precise matches at the Hostname level may match less specific wildcard matches at the Gateway level.
+This allows the Gateway to define a default TLS certificate, which may only be overridden
+
+The HTTPRoute also has a facility to supply additional TLS Config using the `tls` stanza (the RouteTLSConfig field).
+The most important part here is that the field is only used if the `AllowRouteOverride` field is set in the referencing Gateway resource.
+
+Contour will only ever update the `status` of HTTPRoute objects.
+
+Errors or conflicts here will render that section of the config invalid.
```

Comment:
For a missing `serviceName` or `LocalObjectReference`, the docs say: 
```
If the referent cannot be found, the route must be dropped from the Gateway. The controller should raise the “ResolvedRefs” condition on the Gateway with the “DegradedRoutes” reason. The gateway status for this route should be updated with a condition that describes the error more specifically.
```

Does that mean the entire `HTTPRoute` is invalid or just the specific `HTTPRoute.HTTPRouteRule` is invalid?

Reply:
Yeah I wasn't sure about that either (and how it would play with the proposal you had around keeping as much routability as possible even in partially applied resources), I think we chatted about this in office hours or something but I'm not sure there was a conclusion

Reply:
You're right, I think we need to implement this per the spec for now, and raise this upstream as a concern.

Reply:
Chatting in Slack the thought is that only that HTTPRoute.Rule is invalidated so I think the design is correct unless we want to specifically call it out better: https://kubernetes.slack.com/archives/CR0H13KGA/p1612463856030500?thread_ts=1612459180.030400&cid=CR0H13KGA

Reply:
Thanks for asking this question in the upstream Slack @stevesloka. I agree with Harry's answer there - we will implement this as "conflicts will invalidate that *rule*" not "conflicts with invalidate the whole Route" for now, and see if we can push that upstream.